File: app/src/main/java/com/bogdanorzea/popularmovies/ui/main/TopRatedTab.java
package com.bogdanorzea.popularmovies.ui.main;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.bogdanorzea.popularmovies.R;
import com.bogdanorzea.popularmovies.data.MoviesContract;
import com.bogdanorzea.popularmovies.model.object.Movie;
import com.bogdanorzea.popularmovies.model.response.MoviesResponse;
import com.bogdanorzea.popularmovies.utility.AsyncTaskUtils;
import com.bogdanorzea.popularmovies.utility.NetworkUtils;
import com.wang.avi.AVLoadingIndicatorView;

import java.util.ArrayList;

import okhttp3.HttpUrl;

import static com.bogdanorzea.popularmovies.data.MovieMapper.toContentValues;

public class TopRatedTab extends Fragment {

    private PosterAdapter mAdapter;
    private AVLoadingIndicatorView mAvi;
    private TextView warningTextView;
    private boolean isLoading = false;
    private int pageNumber = 1;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        Context context = getContext();

        View view = inflater.inflate(R.layout.layout_recycler_view, container, false);
        mAvi = view.findViewById(R.id.avi);
        warningTextView = view.findViewById(R.id.warning);

        mAdapter = new PosterAdapter(context);

        RecyclerView mRecyclerView = view.findViewById(R.id.recycler_view);
        mRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), 2));
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setAdapter(mAdapter);
        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);

                if (!recyclerView.canScrollVertically(1) && !isLoading) {
                    loadNextPage();
                }
            }
        });

        if (savedInstanceState != null) {
            ArrayList<Movie> movieList = savedInstanceState.getParcelableArrayList("movie_array");
            pageNumber = savedInstanceState.getInt("page_number");
            mAdapter.addMovies(movieList);
        } else {
            loadNextPage();
        }

        return view;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelableArrayList("movie_array", (ArrayList<Movie>) mAdapter.getMovies());
        outState.putInt("page_number", pageNumber);
    }

    private void loadNextPage() {
        if (NetworkUtils.hasInternetConnection(getContext())) {
            HttpUrl url = NetworkUtils.movieTopRatedUrl(pageNumber++);

            AsyncTaskUtils.RequestTaskListener<MoviesResponse> listener =
                    new AsyncTaskUtils.RequestTaskListener<MoviesResponse>() {

                        @Override
                        public void onTaskStarting() {
                            showProgress();
                        }

                        @Override
                        public void onTaskComplete(MoviesResponse moviesResponse) {
                            if (moviesResponse != null) {

                                mAdapter.addMovies(moviesResponse.getResults());

                                for (Movie movie : moviesResponse.getResults()) {
                                    getContext().getContentResolver().insert(MoviesContract.CONTENT_URI, toContentValues(movie));
                                }

                                isLoading = false;
                                hideProgress();
                            }
                        }
                    };

            new AsyncTaskUtils.RequestTask<>(listener, MoviesResponse.class).execute(url);
        } else {
            if (mAdapter.isEmpty()) {
                displayWarning(getString(R.string.warning_no_internet));
            } else {
                Toast.makeText(getContext(), getString(R.string.warning_no_internet), Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void showProgress() {
        isLoading = true;
        mAvi.smoothToShow();
        hideWarning();
    }

    private void hideProgress() {
        isLoading = false;
        mAvi.smoothToHide();
    }

    private void displayWarning(String message) {
        hideProgress();
        warningTextView.setVisibility(View.VISIBLE);
        warningTextView.setText(message);
    }

    private void hideWarning() {
        warningTextView.setVisibility(View.GONE);
    }

}

File: app/src/main/java/com/bogdanorzea/popularmovies/ui/main/TopRatedTab.kt
package com.bogdanorzea.popularmovies.ui.main

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.bogdanorzea.popularmovies.R
import com.bogdanorzea.popularmovies.data.MoviesContract
import com.bogdanorzea.popularmovies.data.toContentValues
import com.bogdanorzea.popularmovies.model.`object`.Movie
import com.bogdanorzea.popularmovies.model.response.MoviesResponse
import com.bogdanorzea.popularmovies.utility.NetworkUtils
import com.bogdanorzea.popularmovies.utility.hasInternetConnection
import kotlinx.android.synthetic.main.layout_recycler_view.*
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch
import java.util.*

class TopRatedTab : Fragment() {

    private lateinit var adapter: PosterAdapter
    private var isLoading = false
    private var pageNumber = 1

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.layout_recycler_view, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        adapter = PosterAdapter(context!!)

        recyclerView.layoutManager = GridLayoutManager(context, 2)
        recyclerView.setHasFixedSize(true)
        recyclerView.adapter = adapter
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)

                if (!recyclerView!!.canScrollVertically(1) && !isLoading) {
                    loadNextPage()
                }
            }
        })

        if (savedInstanceState != null) {
            val movieList =
                    savedInstanceState.getParcelableArrayList<Movie>("movie_array").toList()
            pageNumber = savedInstanceState.getInt("page_number")
            adapter.addMovies(movieList)
        } else {
            loadNextPage()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList("movie_array", adapter.movies as ArrayList<Movie>)
        outState.putInt("page_number", pageNumber)
    }

    private fun loadNextPage() {
        if (context?.hasInternetConnection() == true) {
            launch(UI) {
                showProgress()
                val response = async(CommonPool) {
                    NetworkUtils.fetchResponse(
                            NetworkUtils.movieTopRatedUrl(pageNumber++),
                            MoviesResponse::class.java)
                }.await()

                response?.results?.let {
                    adapter.addMovies(it)

                    it.forEach {
                        this@TopRatedTab.context?.contentResolver?.insert(MoviesContract.CONTENT_URI, it.toContentValues())
                    }
                }

                isLoading = false
                hideProgress()
            }

        } else {
            if (adapter.isEmpty()) {
                displayWarning(getString(R.string.warning_no_internet))
            } else {
                Toast.makeText(context, getString(R.string.warning_no_internet), Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun showProgress() {
        isLoading = true
        avLoadingIndicator?.smoothToShow()
        hideWarning()
    }

    private fun hideProgress() {
        isLoading = false
        avLoadingIndicator?.smoothToHide()
    }

    private fun displayWarning(message: String) {
        hideProgress()
        warningTextView.visibility = View.VISIBLE
        warningTextView.text = message
    }

    private fun hideWarning() {
        warningTextView.visibility = View.GONE
    }

}
